crates/ruff/src/jupyter/jupyter.rs

```diff
@@ -0,0 +1,204 @@
+use std::fs::File;
+use std::io::{BufReader, BufWriter};
+use std::iter;
+use std::path::Path;
+
+use anyhow::{bail, Context, Result};
+use log::debug;
+use serde::Serialize;
+use serde_json::error::Category;
+
+use crate::jupyter::{CellType, JupyterNotebook, SourceValue};
+
+pub const JUPYTER_NOTEBOOK_EXT: &str = "ipynb";
+
+/// Jupyter Notebook indexing table
+///
+/// When we lint a jupyter notebook, we have to translate the row/column based [Location]
+/// to jupyter notebook cell/row/column.
+#[derive(Debug)]
+pub struct JupyterIndex {
+    /// Enter a row (1-based), get back the cell (1-based)
+    pub row_to_cell: Vec<usize>,
+    /// Enter a row (1-based), get back the cell (1-based)
+    pub row_to_row_in_cell: Vec<usize>,
+}
+
+/// Return `true` if the [`Path`] appears to be that of a jupyter notebook file (`.ipynb`).
+pub fn is_jupyter_notebook(path: &Path) -> bool {
+    path.extension()
+        .map_or(false, |ext| ext == JUPYTER_NOTEBOOK_EXT)
+}
+
+pub fn read_jupyter_notebook(path: &Path) -> Result<Option<JupyterNotebook>> {
+    let notebook: JupyterNotebook = match serde_json::from_reader(BufReader::new(File::open(path)?))
+    {
+        Ok(notebook) => notebook,
+        Err(err) => match err.classify() {
+            Category::Io => return Err(err.into()),
+            Category::Syntax | Category::Eof => {
+                // Maybe someone saved the python sources (those with the `# %%` separator)
+                // as jupyter notebook instead. Let's help them.
+                let contents = std::fs::read_to_string(path)?;
+                // Check if tokenizing was successful and the file is non-empty
+                if !(ruff_rustpython::tokenize(&contents))
+                    .last()
+                    .map_or(true, |last| last.is_err())
+                {
+                    return Err(err).context(format!(
+                        "Expected a jupyter notebook (.{} extension), \
+                        which must be internally stored as json, \
+                        but found a python source file",
+                        JUPYTER_NOTEBOOK_EXT
+                    ));
+                }
+
+                return Err(err).context(format!(
+                    "A jupyter notebook (.{}) must internally be json, \
```

Comment:
Nit: IMO we should capitalize as `Jupyter Notebook`, `JSON`, and `Python` for any user-facing text. (For "correctness", but also for consistency with other user-facing text.)